hail/src/main/scala/is/hail/annotations/aggregators/RegionValuePrevNonnullAnnotationAggregator.scala

```diff
@@ -0,0 +1,38 @@
+package is.hail.annotations.aggregators
+
+import is.hail.annotations.{Annotation, Region, RegionValueBuilder, SafeRow}
+import is.hail.expr.types.virtual.Type
+
+class RegionValuePrevNonnullAnnotationAggregator(t: Type) extends RegionValueAggregator {
+  var last: Annotation = null
+
+  def seqOp(region: Region, offset: Long, missing: Boolean) {
+    if (!missing)
+      last = SafeRow.read(t.physicalType, region, offset)
+  }
+
+  def combOp(agg2: RegionValueAggregator) {
+    val that = agg2.asInstanceOf[RegionValuePrevNonnullAnnotationAggregator]
+    if (that.last != null)
+      last = that.last
```

Comment:
I think you're right - the partition index of each corresponding item is required to disambiguate.

Reply:
After a minute of thought, I *believe* that in our current system, this code is correct for the densify use case (inside a scan in an annotate), but that could change if we change how scans utilize Spark actions,.

Reply:
This aggregator is somewhat unusual in that it is non-commutative, like collect.  For non-commutative aggregators to be well-defined, you need to aggregate over an ordered collection of values (rather than a bag/unordered collection of values).  `agg2` corresponds to the aggregation of a block of elements that appears immediately after the elements that were aggregated into `this`.

Thus, I believe this rule is correct: if the later aggregation has a non-null value, we take that, otherwise we take whatever value was aggregated previously.

I think there is an outstanding bug in ContextRDD.aggregate for non-commutative aggregators that will apply them out of order.  That shouldn't be a problem here (we're doing a scan not an aggregation), but it should get fixed.

You'd have to have the partition index of the aggregator as part of the state.  That would also solve collect (but be messy), it's probably easier to just make sure you scan linearly.